Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp

```diff
@@ -135,31 +139,101 @@ class AcceleratedBufferPool {
+            RELEASE_ASSERT(buffer->surface());
+            GraphicsContextSkia context(sk_ref_sp(buffer->surface()));
+
+            DisplayList::Replayer replayer(context, *displayList.get());
+            replayer.replay(initialClip);
+
+            buffer->completePainting();
+        });
+    }
+
+    static unsigned numPaintingThreads()
+    {
+        unsigned numThreads = std::max(1, std::min(8, WTF::numberOfProcessorCores() / 2));
+        if (const char* numThreadsEnv = getenv("WEBKIT_SKIA_CPU_PAINTING_THREADS")) {
+            auto newValue = parseInteger<unsigned>(StringView::fromLatin1(numThreadsEnv));
+            if (newValue && *newValue <= 8)
+                numThreads = *newValue;
+            else
+                WTFLogAlways("The number of Skia/CPU painting threads is not between 0 and 8. Using the default value %u\n", numThreads);
+        }
+
+        return numThreads;
+    }
+
+private:
+    UnacceleratedThreadedRenderingPool(unsigned numThreads)
+        : m_workerPool(WorkerPool::create("SkiaPaintingThread"_s, numThreads))
+    {
+    }
+
+    ~UnacceleratedThreadedRenderingPool() = default;
+
+    Ref<WorkerPool> m_workerPool;
+};
+
 Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect, const IntRect& mappedTileRect, float contentsScale)
 {
-    auto paintBuffer = [&](Nicosia::Buffer& buffer) {
-        buffer.beginPainting();
+    IntRect initialClip(IntPoint::zero(), tileRect.size());
 
-        GraphicsContextSkia context(sk_ref_sp(buffer.surface()));
-        context.clip(IntRect { IntPoint::zero(), tileRect.size() });
+    auto paintIntoGraphicsContext = [&](GraphicsContext& context) {
+        context.clip(initialClip);
 
         if (!contentsOpaque()) {
             context.setCompositeOperation(CompositeOperator::Copy);
-            context.fillRect(IntRect { IntPoint::zero(), tileRect.size() }, Color::transparentBlack);
+            context.fillRect(initialClip, Color::transparentBlack);
             context.setCompositeOperation(CompositeOperator::SourceOver);
         }
 
         context.translate(-tileRect.x(), -tileRect.y());
         context.scale({ contentsScale, contentsScale });
         paintGraphicsLayerContents(context, mappedTileRect);
+    };
+
+    auto paintBuffer = [&](Nicosia::Buffer& buffer) {
+        buffer.beginPainting();
+
+        GraphicsContextSkia context(sk_ref_sp(buffer.surface()));
+        paintIntoGraphicsContext(context);
 
         buffer.completePainting();
     };
 
     static const char* enableCPURendering = getenv("WEBKIT_SKIA_ENABLE_CPU_RENDERING");
     if (enableCPURendering && strcmp(enableCPURendering, "0")) {
         auto buffer = Nicosia::UnacceleratedBuffer::create(tileRect.size(), contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha);
-        paintBuffer(buffer.get());
+
+        if (auto& threadedRenderingPool = UnacceleratedThreadedRenderingPool::singleton(); threadedRenderingPool.numPaintingThreads()) {
```

Comment:
I think it would be better to calculate the number of threads once here, and only create the worker pool if it's > 0

Reply:
Agreed, I will introduce a helper class to retrieve & cache the env vars.